enterprise_access/apps/content_assignments/api.py

```diff
@@ -21,28 +21,161 @@ def get_assignment_configuration(uuid):
 
 def get_assignments_for_configuration(
     assignment_configuration,
-    state=LearnerContentAssignmentStateChoices.ALLOCATED,
+    **additional_filters,
 ):
     """
     Returns a queryset of all ``LearnerContentAssignment`` records
-    for the given assignment configuration.
+    for the given assignment configuration, optionally filtered
+    further by the provided ``additional_filters``.
     """
     queryset = LearnerContentAssignment.objects.select_related(
         'assignment_configuration',
     ).filter(
         assignment_configuration=assignment_configuration,
-        state=state,
+        **additional_filters,
     )
     return queryset
 
 
+def get_assignments_by_learner_email_and_content(
+    assignment_configuration,
+    learner_emails,
+    content_key,
+):
+    """
+    Returns a queryset of all ``LearnerContentAssignment`` records
+    in the given assignment configuration for the provided list
+    of learner_emails and the given content_key.
+    """
+    return get_assignments_for_configuration(
+        assignment_configuration,
+        learner_email__in=learner_emails,
+        content_key=content_key,
+    )
+
+
 def get_allocated_quantity_for_configuration(assignment_configuration):
     """
     Returns a float representing the total quantity, in USD cents, currently allocated
     via Assignments for the given configuration.
     """
-    assignments_queryset = get_assignments_for_configuration(assignment_configuration)
+    assignments_queryset = get_assignments_for_configuration(
+        assignment_configuration,
+        state=LearnerContentAssignmentStateChoices.ALLOCATED,
+    )
     aggregate = assignments_queryset.aggregate(
         total_quantity=Sum('content_quantity'),
     )
     return aggregate['total_quantity']
+
+
+def allocate_assignments(assignment_configuration, learner_emails, content_key, content_price_cents):
+    """
+    Creates or updates an allocated assignment record
+    for the given ``content_key`` in the given ``assignment_configuration``,
+    for each email in the list of ``learner_emails``.
+
+    For existing assignment records with a (config, learner, content) combination, this function
+    does the following:
+      * If the existing record is cancelled or errored, update the existing record state to allocated
+      * If the existing record is allocated or accepted, don't do anything with the record
```

Comment:
Non-blocking product question to ponder: if an existing record is allocated, should a notification be re-sent?